### PR TITLE
chore:  optimize metadata querying performance

### PR DESCRIPTION
**Pull request by dengmingtong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

1, Refactor metadata to optimize performance.
2, Add scan metadata workflow arn to stack output
3, Support manually trigger scan metadata workflow. If workflow is triggerred manually, will not update workflow info.
4, Update interval of checking scan metadata sql from 30 seconds to 60 seconds

Add Unit to cover changes

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [x] all code changes are covered by unit tests
- [x] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [x] deploy data modeling
    - [x] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/analytics/private/sqls/redshift/sp-scan-metadata.sql [hunk 1]
```diff
@@ -278,70 +280,54 @@ BEGIN
...
-							properties_temp_table
-						WHERE 
-							property_value IS NOT NULL AND
-							property_value != ''
-						GROUP BY event_name, project_id, app_info_app_id, property_category, month, day_number, property_name, property_value, value_type, platform									
-					)
-					GROUP BY event_name, project_id, app_info_app_id, property_category, month, day_number, property_name, property_value, value_type
+						value_type, 
+						platform,
+						count(*) AS parameter_count
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-scan-metadata.sql:311]: Try to use `HyperLogLog` to optimize the performance in large dataset

https://docs.aws.amazon.com/redshift/latest/dg/hyperloglog-overview.html

File: src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts [hunk 2]
```diff
@@ -262,10 +165,105 @@ async function batchWriteIntoDDB(metadataItems: any[]) {
         [ddbTableName]: itemsChunk,
       },
     };
-    await ddbDocClient.send(new BatchWriteCommand(inputPara));
+    try {
+      await ddbDocClient.send(new BatchWriteCommand(inputPara));
+    } catch (error) {
+      // log error and inputPara
+      if (error instanceof Error) {
+        logger.error('Error when batch write into ddb.', error);
```
Comment (zxkane) [inline src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts:173]: Log as a single message with an object with multiple parameters

File: src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts [hunk 3]
```diff
@@ -262,10 +165,105 @@ async function batchWriteIntoDDB(metadataItems: any[]) {
...
+    } else {
+      monthValue = checkLatestMonthWithMemory(memoryItemMap, id, currentMonth, markedLatestMonthMap);
+    }
+    if (monthValue === 'latest') {
+      // update memory latest month
+      markedLatestMonthMap.set(id, currentMonth);
+    }
+    return monthValue;
+  } catch (error) {
+    if (error instanceof Error) {
```
Comment (zxkane) [inline src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts:213]: It's not necessary to check the error type.

**Review by zxkane — CHANGES_REQUESTED** [review]
@tyyzqmf please make the corresponding changes for the backend and frontend

File: src/analytics/private/sqls/redshift/sp-scan-metadata.sql (hunk 1 above)
Comment (dengmingtong) [reply src/analytics/private/sqls/redshift/sp-scan-metadata.sql:311]: Currently, just use group by to count parameter count. The performance is fine.

File: src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts (hunk 3 above)
Comment (dengmingtong) [reply src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts]: Refactored

File: src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts (hunk 2 above)
Comment (dengmingtong) [reply src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts]: Refactored it.

File: frontend/src/ts/utils.ts [hunk 4]
```diff
@@ -382,12 +382,21 @@ export const defaultGenericsValue = <T>(expectValue: T, defaultValue: T) => {
...
 
 export const getEventParameters = (
+  metadataEventParameters: IMetadataEventParameter[],
   metadataEvents: IMetadataEvent[],
   eventName?: string
 ) => {
   if (!eventName) {
     return [];
   }
+  if (
```
Comment (chenhaiyun) [inline frontend/src/ts/utils.ts:392]: Maybe can change to `metadataEventParameters?.[0].eventNames?.length > 0` ?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/lambda/api/config/dictionary.json [hunk 5]
```diff
@@ -454,35 +440,7 @@
...
-          "displayName": {
-            "en-US": "User engagement duration(msec)",
-            "zh-CN": "用户互动时长(毫秒）"
-          },
-          "description": {
-            "en-US": "User engagement duration in milliseconds on current screen or page",
-            "zh-CN": "用户在当前界面或页面的互动时长的毫秒数"
-          },
-          "eventName": "_user_engagement"
+          "eventNames": ["_screen_view", "_page_view", "_user_engagement"]
```
[the remark this answers is not in the document]
Comment (zhu-xiaowei) [reply src/control-plane/backend/lambda/api/config/dictionary.json:443]: Yes, the same naming attribute in different event may have different meanings, at here `_engagement_time_msec` in screen view means "previous screen engagement time(msec)"

[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/config/dictionary.json]: support same parameter name in different event

File: src/control-plane/backend/lambda/api/common/utils.ts [hunk 6]
```diff
@@ -1141,6 +1250,11 @@ function _getRunningStatus(lastAction: string) {
       break;
   }
   return status;
+};
+
+function isNewMetadataVersion(pipeline: IPipeline) {
+  const oldVersions = ['v1.0.0', 'v1.0.1', 'v1.0.2', 'v1.0.3', 'v1.1.0', 'v1.1.1'];
```
[the remark this answers is not in the document]
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/common/utils.ts]: Redefine:
 ```
 const unSupportVersions = ['v1.0.0', 'v1.0.1', 'v1.0.2', 'v1.0.3'];
  const oldVersions = ['v1.1.0', 'v1.1.1'];
```

File: src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts [hunk 7]
```diff
@@ -67,7 +68,7 @@ export const handler = async (event: StoreMetadataEvent) => {
     };
   } catch (err) {
     if (err instanceof Error) {
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts:70]: fixed

File: src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts [hunk 8]
```diff
@@ -262,10 +165,100 @@ async function batchWriteIntoDDB(metadataItems: any[]) {
         [ddbTableName]: itemsChunk,
       },
     };
-    await ddbDocClient.send(new BatchWriteCommand(inputPara));
+    try {
+      await ddbDocClient.send(new BatchWriteCommand(inputPara));
+    } catch (error) {
+      // log error and inputPara
+      logger.error('Error when batch write into ddb: ', { error, inputPara });
```
[the remark this answers is not in the document]
Comment (dengmingtong) [reply src/analytics/lambdas/scan-metadata-workflow/store-metadata-into-ddb.ts:170]: Add throw

**Review by zxkane — APPROVED** [review]

**Review by zxkane — APPROVED** [review]

File: frontend/src/ts/utils.ts [hunk 9]
```diff
@@ -382,19 +383,58 @@ export const defaultGenericsValue = <T>(expectValue: T, defaultValue: T) => {
...
 export const getEventParameters = (
+  metadataEventParameters: IMetadataEventParameter[],
   metadataEvents: IMetadataEvent[],
+  builtInMetadata?: IMetadataBuiltInList,
   eventName?: string
 ) => {
   if (!eventName) {
     return [];
   }
+  if (metadataEventParameters?.[0].eventNames?.length > 0) {
```
Comment (chenhaiyun) [inline frontend/src/ts/utils.ts:394]: missing ? after [0], should be `[0]?.eventNames`

File: frontend/src/pages/analytics/metadata/user-attributes/MetadataUserAttributeSplitPanel.tsx [hunk 10]
```diff
@@ -182,6 +189,18 @@ const MetadataUserAttributeSplitPanel: React.FC<
...
                             onClick={() => {
+                              if (
+                                attributeDetails.metadataSource ===
+                                MetadataSource.PRESET
+                              ) {
+                                alertMsg(
+                                  t(
+                                    'analytics:valid.metadataNotAllowEditError'
+                                  ),
+                                  COMMON_ALERT_TYPE.Error as AlertType
```
Comment (chenhaiyun) [inline frontend/src/pages/analytics/metadata/user-attributes/MetadataUserAttributeSplitPanel.tsx:200]: 1. set alertMsg default alertType in `utils.ts`
```
export const alertMsg = (alertTxt: string, alertType="error") => {
```
2. remove COMMON_ALERT_TYPE, replace it with 'error' | 'warning' | 'info' | 'success';

3. optimize this in all source code.

**Pull request merged**